pkg/agent/docker_secrets_postrenderer.go

```diff
@@ -54,3 +76,105 @@ func (r *DockerSecretsPostRenderer) Run(renderedManifests *bytes.Buffer) (modifi
 
 	return modifiedManifests, nil
 }
+
+// updatePodSpecWithPullSecrets updates the podSpec inline with the relevant pull secrets.
+// We do not parse the yaml into actual Kubernetes objects since we want to be
+// independent of api versions. This requires special care and limitations, so
+// we limit our assumptions of the untyped handling to the following, with
+// clear errors when the assumption fails:
+// - The pod spec includes a 'containers' key with a slice value (with a clear error if this is not the case)
+// - Each container value is a map with an 'image' key and string value.
+func (r *DockerSecretsPostRenderer) updatePodSpecWithPullSecrets(podSpec map[interface{}]interface{}) error {
+	containersObject, ok := podSpec["containers"]
+	if !ok {
+		return fmt.Errorf("pod spec did not include a containers key: %+v", podSpec)
+	}
+	containers, ok := containersObject.([]interface{})
```

Comment:
> there is a package in `client-go` that can help with this kind of conversions and parameters retrieval: https://godoc.org/github.com/contiv/client-go/pkg/apis/meta/v1/unstructured

You meant to link to the one owned by kubernetes in apimachinery (rather than `contiv`) right? https://godoc.org/github.com/kubernetes/apimachinery/pkg/apis/meta/v1/unstructured

If Helm's API passed a JSON stream, rather than a YAML stream, I could use that nicely. But as it is, I'd either need to convert the YAML to JSON to parse into the `unstructured.Unstructured`, update, then marshal it out to JSON (using `unstructured` and then back to YAML, or explicitly convert the `map[interface{}]interface{}` to `map[string]interface{}` to be able to use `unstructured.Unstructured.SetUnstructuredContent()` to create the object in the first place. Let me know if I've missed something - happy to try.

> 
> Also, I would try to use the actual struct types of the resources we are handling instead keep using type conversions from `interface{}` and literal keys. It's easier to just access `podSpec.Containers[0].Image` rather than doing several conversions.

Yes, I had gone this way initially for the same reason, but realized that we'd then be making assumptions about the templated YAML that are not ours to make (ie. which client-go version we're using affects the resources that we import). IMO we should be heading the other way - ideally we would not change the rest of the YAML at all (no changes in ordering, no re-rendering except the addition of the image pull secret), but to do that is a little more complex.

> It's even more useful when we add more complex types like a `CronJob` in which I don't necessarily know the full path till the image name.

They're actually pretty trivial - they just have a `spec.template.spec` as do all the others (I left a comment linking to https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#podtemplatespec-v1-core ).